### PR TITLE
Added Satellite exercise

### PR DESCRIPTION
**Pull request by tofische**
Added a new Haskell practice exercise Satellite based on the available problem specification.

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -500,6 +500,17 @@
...
+      {
+        "slug": "satellite",
+        "name": "Satellite",
+        "uuid": "22136812-8160-49f0-8a0d-f76a2086a34c",
+        "practices": [],
+        "prerequisites": [],
+        "difficulty": 6,
+        "topics": [
+          "maybe"
+        ]
```
Comment (MatthijsBlom) [inline config.json:512]: @ErikSchierboom I am not familiar with this feature, so I cannot judge this.

File: exercises/practice/satellite/src/Satellite.hs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module Satellite (treeFromTraversals) where
+
+import Data.Tree (Tree)
+
+treeFromTraversals :: String -> String -> Maybe (Tree Char)
```
Comment (MatthijsBlom) [inline exercises/practice/satellite/src/Satellite.hs:5]:
```suggestion
treeFromTraversals :: Ord a => [a] -> [a] -> Maybe (Tree a)
```

Or is this unworkable?

Comment (tofische) [reply exercises/practice/satellite/src/Satellite.hs:5]: Good proposal, done.

**Comment by MatthijsBlom** [comment]
Hmm. This exercise is about binary trees, but `Data.Tree.Tree` is a rose tree. Should we be more precise and use some dedicated `BinaryTree` type instead?

**Comment by tofische** [comment]
> Hmm. This exercise is about binary trees, but `Data.Tree.Tree` is a rose tree. Should we be more precise and use some dedicated `BinaryTree` type instead?

I agree with you, but I couldn't find a ready-to-use binary tree in any common package. I didn't want to use some exotic package, and I also didn't want to let the students to create their own tree implementation.

The closest I could find is [https://hackage.haskell.org/package/tree-traversals-0.1.2.0/docs/Data-BinaryTree.html](Data.BinaryTree) - what do you think about the usage of this package?

**Comment by MatthijsBlom** [comment]
Neat!

If we add a pre-defined type in this way we should probably add a note about it in `instructions.append.md`.

```haskell
-- Which type to choose?

data BinaryTree a = Leaf    | Node a (BinaryTree a) (BinaryTree a)
-- or
data BinaryTree a = Leaf a | Node a (BinaryTree a) (BinaryTree a)
```

**Comment by tofische** [comment]
> Neat!
> 
> If we add a pre-defined type in this way we should probably add a note about it in `instructions.append.md`.
> 
> ```haskell
> -- Which type to choose?
> 
> data BinaryTree a = Leaf   | Node a (BinaryTree a) (BinaryTree a)
> -- or
> data BinaryTree a = Leaf a | Node a (BinaryTree a) (BinaryTree a)
> ```

I have to admit that I'm not keen on the idea of providing a custom-made basic data structure - one of the purposes of the exercises is to point people towards suitable existing libraries. At least for me it was a big benefit that I got from the Exercism, so I don't want to withhold it from others.

What are the arguments against using `Data.BinaryTree`?

**Comment by MatthijsBlom** [comment]
> What are the arguments against using `Data.BinaryTree?`

To me: that it is not [small](https://www.parsonsmatt.org/2018/10/02/small_types.html). Many rose trees are not binary trees. Haskell allows and encourages expressing one's intent precisely. Defining new types even for single use is not unusual.

**Comment by tofische** [comment]
> > What are the arguments against using `Data.BinaryTree?`
> 
> To me: that it is not [small](https://www.parsonsmatt.org/2018/10/02/small_types.html). Many rose trees are not binary trees. Haskell allows and encourages expressing one's intent precisely. Defining new types even for single use is not unusual.

Are we talking about the same thing? Source of `Data.BinaryTree`:

```
data BinaryTree a = Leaf | Branch (BinaryTree a) a (BinaryTree a)
  deriving (Show, Functor, Eq)
```

**Comment by MatthijsBlom** [comment]
No, sorry. I thought we were talking about `Data.Tree` from `containers`.

`tree-traversals` seems not a commonly used package. In any case it would have to be added to the test runner. I have no definite opinion on whether we should use it.

**Comment by tofische** [comment]
> No, sorry. I thought we were talking about `Data.Tree` from `containers`.
> 
> `tree-traversals` seems not a commonly used package. In any case it would have to be added to the test runner. I have no definite opinion on whether we should use it.

This is a valid argument.

File: exercises/practice/satellite/package.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+name: satellite
+version: 1.0.0.0
+
+dependencies:
+  - base
+  - containers
```
Comment (MatthijsBlom) [inline exercises/practice/satellite/package.yaml:6]: `containers` isn't needed anymore.

**Review by MatthijsBlom — APPROVED** [review]

File: exercises/practice/satellite/package.yaml (hunk 3 above)
Comment (tofische) [reply exercises/practice/satellite/package.yaml]: You are correct, it is need only in the example solution which uses Data.Set.

**Comment by ErikSchierboom** [comment]
> tree-traversals seems not a commonly used package. In any case it would have to be added to the test runner. I have no definite opinion on whether we should use it.

When possible, we'd prefer not to add libraries to the test runner. 

>  one of the purposes of the exercises is to point people towards suitable existing libraries

I'd argue that this isn't generally true, although it can be in individual cases. Our idea is that students solve things _without_ using libraries, as Exercism is all about teaching fluency in a language, but not necessarily everything that goes with it (like tooling and libraries).

File: config.json (hunk 1 above)
Comment (ErikSchierboom) [reply config.json:512]: Topics are deprecated and are not used by the website. They are allowed to be in the `config.json` as they used to be used and they could be useful to help determine the `practices` and `prerequisites` fields.

**Review by ErikSchierboom — APPROVED** [review]
LGTM!

**Pull request merged**